### PR TITLE
Generation labeling for regions and allocation requests.

### PR DESCRIPTION
**Pull request by bernd-aws**
The original generational Shenandoah patch authored by Roman Kennke. 
The rest of the project will build upon this base.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ❌ (1/9 failed) | ⏳ (5/9 running) | ⏳ (1/9 running) |

**Failed test task**
- [Linux x64 (hs/tier1 gc)](https://github.com/bernd-aws/shenandoah/runs/1407557017)

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to b90f4dc99096cb0ed2821b345fa28a41f961949e

### Download
`$ git fetch https://git.openjdk.java.net/shenandoah pull/6/head:pull/6`
`$ git checkout pull/6`

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp [hunk 1]
```diff
@@ -261,6 +261,19 @@ inline oop ShenandoahHeap::evacuate_object(oop p, Thread* thread) {
   assert(ShenandoahThreadLocalData::is_evac_allowed(thread), "must be enclosed in oom-evac scope");
 
   size_t size = p->size();
+  ShenandoahGeneration target_gen = heap_region_containing(p)->generation();
+  if (target_gen == YOUNG_GEN) {
+    markWord mark = p->mark();
+    if (mark.is_marked()) {
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp:267]: This could probably use ShenandoahForwarding::is_marked() instead.

**Review by rkennke — APPROVED** [review]
One small thing (and I leave it to you if you want to change it), otherwise looks good.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp (hunk 1 above)
Comment (bernd-aws) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp]: Fixed.

**Comment by rkennke** [comment]
Thanks! I already approved it, you can /integrate it now :-)

**Pull request closed without merging**